File: src/main/java/leetcode/_3/Solution.java
package leetcode._3;

// Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

import java.util.BitSet;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        var marker = new BitSet(127);
        int max = 0, low = 0;

        for (var i = 0; i < s.length(); ) {
            if (!marker.get(s.charAt(i))) {
                marker.set(s.charAt(i));
                i++;
                continue;
            }

            max = Integer.max(max, i - low);

            while (s.charAt(low) != s.charAt(i)) {
                marker.set(s.charAt(low), false);
                low++;
            }

            marker.set(s.charAt(low), false);
            low++;
        }

        return Integer.max(max, s.length() - low);
    }
}

File: src/main/java/leetcode/_3/solution.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	offsets := make([]int, 128)

	longest := 0
	low := 0

	for i := 0; i < len(s); i++ {
		low = max(low, offsets[s[i]])
		longest = max(longest, i-low+1)
		offsets[s[i]] = i + 1
	}

	return longest
}

func max(a int, b int) int {
	if a > b {
		return a
	}

	return b
}

// ***************** Testing *****************
func main() {
	s := "abcabcbb"
	size := lengthOfLongestSubstring(s)
	fmt.Println(size)
}
